**Note to grader:** Each question consists of parts, e.g. Q1(i), Q1(ii), etc. Each part must be graded  on a 0-4 scale, following the standard NJIT convention (A:4, B+: 3.5, B:3, C+: 2.5, C: 2, D:1, F:0). 
The total score must be re-scaled to 100 -- that should apply to all future assignments so that Canvas assigns the same weight on all assignments. 



In [1]:
# Grader's area
import numpy as np
M = np.zeros([10,10])
maxScore = 0


# **Assignment 4**

The goal of this assignment is to run some experiments with scikit-learn on a fairly sizeable and interesting image data set. This is the MNIST data set that consists of lots of images, each having 28x28 pixels. By today's standards, this may seem relatively tiny, but only a few years ago was quite challenging computationally, and it motivated the development of several ML algorithms and models that are now state-of-the-art  solutions for much bigger data sets. 

The assignment is experimental. We will try to whether a combination of PCA and kNN can yield any good results for the MNIST data set. Let's see if it can be made to work on this data set. 

Note: There are less difficult Python parts in this assignment. You can get things done by just repeating things from the class notebooks. But your participation and interaction via Canvas is always appreciated!

## Preparation Steps

In [2]:
# Import all necessary python packages
import numpy as np
import os
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from sklearn.linear_model import LogisticRegression

In [3]:
# we load the data set directly from scikit learn 
# 
# note: this operation may take a few seconds. If for any reason it fails we 
# can revert back to loading from local storage. 

from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split


X, y = fetch_openml('mnist_784', version=1, return_X_y=True)
y = y.astype(int)
X = ((X / 255.) - .5) * 2
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=10000, random_state=123, stratify=y)


## <font color = 'blue'> Question 1. Inspecting the Dataset </font>

**(i)** How many data points are in the training and test sets ? <br>
**(ii)** How many attributes does the data set have ?

Exlain how you found the answer to the first two questions. 

[**Hint**: Use the 'shape' method associated with numpy arrays. ]

**(iii)** How many different labels does this data set have. Can you demonsrate how to read that number from the vector of labels *y_train*?  <br>
**(iv)** How does the number of attributes relates to the size of the images? <br>
**(v)** What is the role of line 12 in the above code? 





*(Please insert cells below for your answers. Clearly id the part of the question you answer)*

In [4]:
# (i) How many data points are in the training and test sets ?
print('Question 1.1')
print("Total train dataset size: ", X_train.shape[0])
print("Total test dataset size: ", X_test.shape[0])

Question 1.1
Total train dataset size:  60000
Total test dataset size:  10000


In [5]:
#(ii) How many attributes does the data set have ?
print('Question 1.2')
print("Total number of attributes: ",X_train.shape[1])

Question 1.2
Total number of attributes:  784


In [6]:
#(iii) How many different labels does this data set have. 
# Can you demonsrate how to read that number from the vector of labels y_train?
print('Question 1.3')
print("Label size: ",np.unique(y_train).size)

Question 1.3
Label size:  10


(iv) How does the number of attributes relates to the size of the images? <br><br>
- There are 784 number of attributes that are classified by pixel size of image.<br>
- Each image size is 28 x 28 which is equal to 784.

(v) What is the role of line 12 in the above code? <br><br>
- The line takes the vector **X** and transofrms the same.<br> 
- We divide the **X** vector by **255** so the values fall in range of **0 and 1**.<br>
- We subtract it from **0.5** and multiply by **2**.<br>
We make all values in the vecot **X** in the range [0,1].<br>
- **X** = ((**X**/**255**)-**0.5**)***2**


In [7]:
# For grader use only

# in this case, make excetion and, assign 0-2 points for each subquestion

# insert grade here  
# G[1,1] = 
# G[1,2] =
# G[1,3] = 
# G[1,4] = 
# G[1,5] =  


maxScore = maxScore + 10


##  <font color = 'blue'> Question 2. PCA on MNIST </font>

Because the number of attributes of the MNIST data set may be too big to apply kNN on it (due to the 'curse of dimensionality'), we want to compress the images down to a smaller number of 'fake' attributes. 

Use scikit-learn to output a data set *X_train_transformed* and *X_test_transformed*, with $l$ attributes. Here a reasonable choice of $l$ is 10, equal to the number of labels. But you can try slightly smaller or bigger values as well. 


**Hint**: Take a look at [this notebook](https://colab.research.google.com/drive/1DG5PjWejo8F7AhozHxj8329SuMtXZ874?usp=drive_fs) we used in the lecture, and imitate what we did there. Be careful though, to use only the scikit-learn demonstration, not the exhaustive PCA steps we did before it.

**Note**: This computation can take a while. If problems are encountered we can try the same experiment on a downsized data set. 

In [8]:
# Question 2 answer:
from sklearn.decomposition import PCA


# transform the data using PCA
n_components = 10
pca = PCA(n_components = n_components)
X_train_transformed = pca.fit_transform(X_train)
X_test_transformed = pca.transform(X_test)
print('Question 2 answered')

Question 2 answered


In [9]:
# for grader use
maxScore = maxScore +4 


# insert grade here (out of 4)
# G[2,1] =



## <font color = 'blue'> Question 3. kNN on MNIST attributes from PCA </font>


Having calculated the *transformed* MNIST data set we can now apply a kNN approach to the MNIST classification data set. Here are the sets:

(i) Fit a $k$-NN classifier on the transformed data set. Here $k$ is a hyperparameter, and you can experiment with it. Be aware though, that larger $k$ can take more time to fit. 

(ii) Apply the classifier on the transformed test set. What is the classification accuracy? 

(iii) A theoretical question: if we skipped all the above steps and we just assigned a **random** label to each test point, what would the classification accuracy be on average?  Does your result (ii) beat the random expectation? 

(iv) Experiment with different settings of $k$, and other hyperparameters that are described in the scikit-learn manual of the kNN classifier. Report your findings in a separate cell. Also for **participation points**: report your best result on Canvas! 

[**Hint**: Imitate the steps from the classroom notebook]


In [10]:
# (i) Fit a  k -NN classifier on the transformed data set. 
# Here  k  is a hyperparameter, and you can experiment with it. 
# Be aware though, that larger  k  can take more time to fit.

from sklearn.neighbors import KNeighborsClassifier
k = 3
knn = KNeighborsClassifier(n_neighbors = k)
knn.fit(X_train_transformed, y_train)

print('Question 3.1')
print('k-NN fit on transformed dataset k = 3')

Question 3.1
k-NN fit on transformed dataset k = 3


In [11]:
# (ii) Apply the classifier on the transformed test set. What is the classification accuracy?
# display the accuracy for various k values

print('Question 3.2')
print('For k =', str(k) + ': Accuracy =', knn.score(X_test_transformed, y_test))

Question 3.2
For k = 3: Accuracy = 0.9322


In [12]:
# (iii) A theoretical question: if we skipped all the above steps and we just assigned 
# a random label to each test point, what would the classification accuracy be on average? 
# Does your result (ii) beat the random expectation?
from sklearn.utils import shuffle
y_test_random = shuffle(y_test)

print('Question 3.3')
print('For k =', str(k) + ' on original test set: Accuracy =', knn.score(X_test_transformed, y_test))
print('For k =', str(k) + ' on random test set: Accuracy =', knn.score(X_test_transformed, y_test_random))
print('The classification accuracy is higher with the labels on the original data set.')

Question 3.3
For k = 3 on original test set: Accuracy = 0.9322
For k = 3 on random test set: Accuracy = 0.0939
The classification accuracy is higher with the labels on the original data set.


In [13]:
#(iv) Experiment with different settings of  k , and other hyperparameters that 
# are described in the scikit-learn manual of the kNN classifier. Report your findings in a separate cell. 
print('Question 3.4')
for k in list(range(1,11)):
  knn = KNeighborsClassifier(n_neighbors=k)
  knn.fit(X_train_transformed, y_train)
  print('For k =', str(k) + ': Accuracy =', knn.score(X_test_transformed, y_test))

Question 3.4
For k = 1: Accuracy = 0.9185
For k = 2: Accuracy = 0.9145
For k = 3: Accuracy = 0.9322
For k = 4: Accuracy = 0.9316
For k = 5: Accuracy = 0.9347
For k = 6: Accuracy = 0.9329
For k = 7: Accuracy = 0.9347
For k = 8: Accuracy = 0.9337
For k = 9: Accuracy = 0.9347
For k = 10: Accuracy = 0.9331


In [14]:
# for grader use
maxScore = maxScore +12

# insert grade here (each item out of 4)
# G[3,1] =
# G[3,2] = 
# G[3,3] =
# G[3,4] = 

In [15]:
# for grader use

# Total Grade Calculation
rawScore = np.sum(G)
score = rawScore*100/maxScore

NameError: ignored